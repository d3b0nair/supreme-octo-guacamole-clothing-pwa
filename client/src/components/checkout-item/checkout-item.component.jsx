import React from "react";
import { connect } from "react-redux";
import CustomButton from "../custom-button/custom-button.component";
import {
  clearItemFromCart,
  addItem,
  removeItem,
} from "../../redux/cart/cart.actions";

import {
  CheckoutItemContainer,
  ImageContainer,
  ImgWrapper,
  CollumnWrapper,
  QuantityCollumn,
  ArrowSymbol,
  ItemValue,
  RemoveButton,
  RemoveButtonWrapper,
} from "./checkout-item.styles";

export function CheckoutItem({ cartItem, clearItem, addItem, removeItem }) {
  const smallScreen = window.innerWidth < 300;
  const { name, imageUrl, price, quantity } = cartItem;
  return (
    <CheckoutItemContainer>
      <ImageContainer>
        <ImgWrapper src={imageUrl} alt="item" />
      </ImageContainer>
      <CollumnWrapper>{name}</CollumnWrapper>
      <QuantityCollumn>
        <ArrowSymbol
          data-testid="remove-checkout-item"
          onClick={() => removeItem(cartItem)}
        >
          &#10094;
        </ArrowSymbol>
        <ItemValue>{quantity}</ItemValue>
        <ArrowSymbol
          data-testid="add-checkout-item"
          onClick={() => addItem(cartItem)}
        >
          &#10095;
        </ArrowSymbol>
      </QuantityCollumn>
      <CollumnWrapper>
        {smallScreen ? `Price: ${price}` : price}$
      </CollumnWrapper>
      <RemoveButtonWrapper>
        <RemoveButton
          data-testid="clear-checkout-item"
          onClick={() => clearItem(cartItem)}
        >
          {smallScreen ? (
            <CustomButton inverted>Remove Item</CustomButton>
          ) : (
            "âœ•"
          )}
        </RemoveButton>
      </RemoveButtonWrapper>
    </CheckoutItemContainer>
  );
}

const mapDispatchToProps = (dispatch) => ({
  clearItem: (item) => dispatch(clearItemFromCart(item)),
  addItem: (item) => dispatch(addItem(item)),
  removeItem: (item) => dispatch(removeItem(item)),
});

export default connect(null, mapDispatchToProps)(CheckoutItem);
