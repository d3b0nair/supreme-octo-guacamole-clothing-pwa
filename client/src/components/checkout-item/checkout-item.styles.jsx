import styled, { css } from "styled-components";

const scaleUpOnActive = css`
  &:active {
    transform: scale(1.65);
    transition: transform 0.6s cubic-bezier(0.25, 0.45, 0.45, 0.95);
  }
`;
const noTextSelectionColor = css`
  &::selection {
    opacity: 0;
  }
`;

const onHover = css`
  &:hover {
    color: #ff6e7f;
  }
`;

const buttonStyle = css`
  cursor: pointer;
  ${scaleUpOnActive}
  ${noTextSelectionColor}
  ${onHover}
`;

export const CheckoutItemContainer = styled.div`
  width: 100%;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr;
  gap: 0px 0px;
  grid-template-areas: ". . . . .";
  min-height: 100px;
  border-bottom: 1px solid white;
  font-size: 20px;
  align-items: center;
  @media screen and (max-width: 300px) {
    display: flex;
    flex-direction: column;
  }
`;

export const ImageContainer = styled.div`
  padding: 5px;
  @media screen and (max-width: 800px) {
    width: auto;
  }
`;

export const ImgWrapper = styled.img`
  max-width: 100%;
  height: auto;
  border: 1px solid rgba(255, 255, 255, 1);
`;
export const CollumnWrapper = styled.span`
  text-align: center;
  display: block;
  font-weight: normal;
`;
export const QuantityCollumn = styled.span`
  display: flex;
  justify-content: center;
`;
export const ArrowSymbol = styled.div`
  ${buttonStyle}
`;
export const ItemValue = styled.div`
  margin: 0 10px;
`;

export const RemoveButton = styled.div`
  ${buttonStyle}
`;
export const RemoveButtonWrapper = styled.div`
  text-align: center;
  display: flex;
  justify-content: center;
`;
