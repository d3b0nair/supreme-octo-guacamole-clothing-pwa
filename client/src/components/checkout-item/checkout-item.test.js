import React from "react";
import {
  render,
  fireEvent,
  waitFor,
  screen,
  cleanup,
} from "@testing-library/react";
import { CheckoutItem } from "./checkout-item.component";

const mockRemoveItem = jest.fn();
const mockAddItem = jest.fn();
const mockClearItem = jest.fn();
const mockProps = (otherProps) => ({
  ...otherProps,
  cartItem: [],
  clearItem: mockClearItem,
  addItem: mockAddItem,
  removeItem: mockRemoveItem,
});
const component = (mockProps) => <CheckoutItem {...mockProps} />;

afterEach(cleanup);

describe("CheckoutItem component", () => {
  it("should mount", () => {
    const { container } = render(component(mockProps()));
    expect(container).toMatchSnapshot();
  });
  it("should fire addItem, removeIten, clearItem", async () => {
    render(component(mockProps()));
    await waitFor(() => fireEvent.click(screen.getByTestId("add-checkout-item")));
    expect(mockAddItem).toHaveBeenCalled();
    await waitFor(() => fireEvent.click(screen.getByTestId("remove-checkout-item")));
    expect(mockRemoveItem).toHaveBeenCalled();
    await waitFor(() => fireEvent.click(screen.getByTestId("clear-checkout-item")));
    expect(mockClearItem).toHaveBeenCalled();
  });
});
