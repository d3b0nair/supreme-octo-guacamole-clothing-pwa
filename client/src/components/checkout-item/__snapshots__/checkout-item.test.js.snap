// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutItem component should mount 1`] = `
<div>
  <div
    class="sc-gsDKAQ ksGHNo"
  >
    <div
      class="sc-dkPtRN gUwQJy"
    >
      <img
        alt="item"
        class="sc-hKwDye gkmiPh"
      />
    </div>
    <span
      class="sc-eCImPb fqRrfq"
    />
    <span
      class="sc-jRQBWg fGFNQO"
    >
      <div
        class="sc-gKclnd lcsmPG"
        data-testid="remove-checkout-item"
      >
        ❮
      </div>
      <div
        class="sc-iCfMLu jHWOzV"
      />
      <div
        class="sc-gKclnd lcsmPG"
        data-testid="add-checkout-item"
      >
        ❯
      </div>
    </span>
    <span
      class="sc-eCImPb fqRrfq"
    >
      $
    </span>
    <div
      class="sc-pVTFL fjkOjk"
    >
      <div
        class="sc-furwcr gxgspo"
        data-testid="clear-checkout-item"
      >
        ✕
      </div>
    </div>
  </div>
</div>
`;
